### PR TITLE
Added list of known dependency compatibilities and regeneration infrastructure for it

### LINKED ISSUE #1751: [BUG]: `PipResolver_locate_dist_info` matches multiple `.dist-info`'s
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

The current logic in `PipResolver_locate_dist_info` selects the first `.dist-info` it finds for a library. However, multiple `dist-info's` could be selected, which migth result in selecting the wrong `.dist-info`:

``` python
import databricks
```

``` sh
> ls .venv/lib/python3.10/site-packages/databricks*
.venv/lib/python3.10/site-packages/databricks
.venv/lib/python3.10/site-packages/databricks_labs_blueprint-0.4.4.dist-info
.venv/lib/python3.10/site-packages/databricks_labs_lsql-0.4.2.dist-info
.venv/lib/python3.10/site-packages/databricks_labs_pylint-0.4.0.dist-info
.venv/lib/python3.10/site-packages/databricks_labs_ucx-0.22.0.dist-info
.venv/lib/python3.10/site-packages/databricks_sdk-0.27.0.dist-info
```

### Expected Behavior

Selecting multiple `.dist-info`'s should be handled

### Steps To Reproduce

1. Create test for above example
2. Handle selecting multiple `.dist-info`'s

### Cloud

Azure

### Operating System

macOS

### Version

latest via Databricks CLI

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by nfx**
## Changes

This PR introduces automated regeneration of known python dependencies for their UC compatibility to terminate import resolution of the graph on them.

- [x] add script entrypoint for adding new libraries
- [x] manually trim `known.json` from overly detailed info for packages that don't work with Spark.
- [x] read `known.json` as package data and integrate it with the Whitelist

Fixes #1751

### CODE REVIEW

File: src/databricks/labs/ucx/source_code/known.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                filename = line.split(',')[0]
+                if not filename.endswith(".py"):
+                    continue
+                files.append(self._path.parent / filename)
+        return files
+
+    @cached_property
+    def _metadata(self):
+        with Path(self._path, "METADATA").open(encoding=_DEFAULT_ENCODING) as f:
+            return email.message_from_file(f)
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/known.py:141]: 👍

**Review by JCZuurmond — CHANGES_REQUESTED** [review]
some implementation details

File: src/databricks/labs/ucx/source_code/known.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+    def distribution_compatibility(self, name: str) -> Compatibility:
+        if not name:
+            return UNKNOWN
+        for module, distribution_name in self._module_distributions.items():
+            if distribution_name != name:
+                continue
+            problems = self._module_problems[module]
+            return Compatibility(True, problems)
+        return UNKNOWN
```
[the remark this answers is not in the document]
Comment (nfx) [reply src/databricks/labs/ucx/source_code/known.py:70]: Distribution is value, not a key

File: src/databricks/labs/ucx/source_code/pyspark.py [hunk 3]
```diff
@@ -40,12 +40,17 @@ def apply(self, from_table: FromTable, index: MigrationIndex, node: ast.Call) ->
         """applies recommendations"""
 
     def _get_table_arg(self, node: ast.Call):
-        if len(node.args) > 0:
-            return node.args[self.table_arg_index] if self.min_args <= len(node.args) <= self.max_args else None
-        if not self.table_arg_name or not node.keywords:
+        node_argc = len(node.args)
```
Comment (ericvergnaud) [inline src/databricks/labs/ucx/source_code/pyspark.py:43]: seems unrelated to this PR

File: src/databricks/labs/ucx/source_code/pyspark.py [hunk 4]
```diff
@@ -187,10 +192,12 @@ def matches(self, node: ast.AST):
 
     def lint(self, from_table: FromTable, index: MigrationIndex, node: ast.Call) -> Iterator[Advice]:
         table_arg = self._get_table_arg(node)
-
         if not isinstance(table_arg, ast.Constant):
             return
-
+        if not table_arg.value:
```
Comment (ericvergnaud) [inline src/databricks/labs/ucx/source_code/pyspark.py:197]: seems unrelated to this PR

**Review by ericvergnaud — COMMENTED** [review]
Populating the WhiteList from UCX dependencies is a good starting point but are we sure it is good enough ?  I suspect there are numerous libraries used by clients that are 100% compatible that UCX knows nothing about ? The previous approach was to rely on a manually populated list. Should we not combine the 2 approaches ?
